contracts/eosio.system/src/finalizer_key.cpp

```diff
@@ -0,0 +1,158 @@
+#include <eosio.system/eosio.system.hpp>
+
+#include <eosio/eosio.hpp>
+#include <eosio/privileged.hpp>
+#include <eosio/crypto_bls_ext.hpp>
+
+namespace eosiosystem {
+
+   bool system_contract::is_savanna_consensus() {
+      return !_gstate5.last_finalizers.empty();
+   }
+
+   bool system_contract::is_registered_producer( const name& producer ) {
+      // Returns true if `producer` is a registered producer
+   }
+
+   bool system_contract::is_active_producer( const name& producer ) {
+      // Returns true if `producer` is an active producer (top 21)
+   }
+
+   bool set_next_finalizers_set(const std::vector<name>& producers ) {
+      // Build finalizer set
+      std::vector<finalizer_authority> finalizers;
+      for ( const auto p&: producers ) {
+         auto it = _finalizer_keys.find( p.value );
+         if( it != _finalizer_keys.end() ) { // p has registered a finalizer key
+            finalizers.emplace_back(
+               finalizer_authority{
+                  .description = "",
+                  .weight      = 1,
+                  .public_key  = it->active_key
+               }
+            );
+         }
+      }
+
+      // Abort if less than 15 top producers have registered a finalizer key
+      if( finalizers.size < 15 ) {
+         return false;
+      }
+
+      // Establish finalizer policy and call set_finalizers() host function
+      finalizer_policy fin_policy {
+         .generation = 0;
+         .threshold  = 15;
+         .finalizers = finalizers;
+      };
+      set_finalizers(std::move(fin_policy));
```

Comment:
There needs to be a comparison of the constructed `finalizers` to the one that was previously stored in `_gstate5.last_finalizers` to ensure they are not the same. If they are the same we need to avoid calling `set_finalizers`.